controller/api/destination/server_adaptor.go

```diff
@@ -0,0 +1,36 @@
+package destination
+
+import (
+	"github.com/linkerd/linkerd2/controller/api/destination/watcher"
+	"github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha2"
+)
+
+type serverAdaptor struct {
+	listener        watcher.ProfileUpdateListener
+	port            watcher.Port
+	isOpaque        bool
+	annotationPorts map[uint32]struct{}
+}
+
+func newServerAdaptor(listener watcher.ProfileUpdateListener, port watcher.Port) *serverAdaptor {
+	return &serverAdaptor{
+		listener: listener,
+		port:     port,
+	}
+}
+
+func (sa *serverAdaptor) Update(isOpaque bool) {
+	sa.isOpaque = isOpaque
+	ports := make(map[uint32]struct{})
+	for port := range sa.annotationPorts {
+		ports[port] = struct{}{}
+	}
+	// If a Server has marked the port as opaque, then ensure the port is in
+	// the set of opaque ports.
+	if isOpaque {
+		ports[sa.port] = struct{}{}
+	}
+	sp := v1alpha2.ServiceProfile{}
```

Comment:
outside the context of the GetProfile implementation, it's not clear why this returns a blank serviceprofile.  I think this needs some documentation explaining that it's only to be used in the context of a GetProfile call directly on an endpoint, meaning that there will never be a real service profile and that this is purely to create a synthetic service profile purely to attach opaque port info.  otherwise, it's easy to assume that this is similar to other adapters and that it adapts existing service profiles with server information or something like that.